lib/Ptero/Concrete/Workflow.pm

```diff
@@ -1,149 +1,75 @@
 package Ptero::Concrete::Workflow;
 
-use Moose;
+use strict;
 use warnings FATAL => 'all';
-
-use JSON qw();
-use Params::Validate qw(validate_pos :types);
-
-use Ptero::Concrete::DAG;
-
-with 'Ptero::Builder::Detail::HasWebhooks';
-with 'Ptero::Concrete::Detail::Roles::CanWriteReport';
-
-my $codec = JSON->new()->canonical([1]);
-
-has 'dag' => (
-    is => 'ro',
-    isa => 'Ptero::Concrete::DAG',
-);
-
-has 'inputs' => (
-    is => 'ro',
-    isa => 'HashRef',
-);
-
-has 'name' => (
-    is => 'ro',
-    isa => 'Str',
-);
-
-has 'status' => (
-    is => 'ro',
-    isa => 'Str',
-);
-
-has 'url' => (
-    is => 'ro',
-    isa => 'Str',
-);
-
-sub write_report {
-    my $self = shift;
-    my %p = Params::Validate::validate(@_, {
-        handle => 1,
-        indent => {default => 0},
-    });
-
-    $self->_write_report($p{handle}, $p{indent}, 0);
-}
-
-sub _write_report {
-    my $self = shift;
-    my ($handle, $indent, $color) = $self->params_validator(@_);
-
-    printf $handle $self->format_line,
-        'TYPE',
-        'STATUS',
-        'STARTED',
-        'DURATION',
-        'P-INDEX',
-        '',
-        'NAME';
-
-    printf $handle $self->format_line,
-        'Workflow',
-        $self->status,
-        '',
-        '',
-        '',
-        $self->indentation_str x $indent,
-        $self->name;
-
-    for my $name ($self->dag->sorted_tasks) {
-        my $task = $self->dag->task_named($name);
-        $task->_write_report($handle, $indent+1, $color);
+use Params::Validate qw(validate validate_pos :types);
+use IO::String;
+
+use Ptero::Concrete::Workflow::Task;
+use Ptero::Concrete::Workflow::Execution;
+use Ptero::Concrete::Workflow::ReportWriter;
+
+sub new {
+    my ($class, $hashref) = @_;
+
+    my $self = {};
+    $self->{id} = $hashref->{id};
+    $self->{root_task_id} = $hashref->{rootTaskId};
+    $self->{name} = $hashref->{name};
+    $self->{status} = $hashref->{status};
+    $self->{method_index} = {};
+    $self->{task_index} = {};
+
+    my $task_data = $hashref->{tasks};
+    for my $task_name (keys %$task_data) {
```

Comment:
Not sure if this would work, but this seems clearer to me:
```
while (($key, $val) = each %{$hashref->{tasks}}) {
  $self->{tasks}{$key} = Ptero::Concrete::Workflow::Task->new($key, $val);
}
```